File: src/main/java/org/dedda/bratwurst/lang/Calculation.java
package org.dedda.bratwurst.lang;

/**
 * Created by dedda on 10/15/15.
 *
 * @author dedda
 */
public class Calculation extends BWExpression {

    private BWExpression leftSide;
    private BWExpression rightSide;
    private char operator;
    private int value = 0;

    public Calculation(BWExpression leftSide, BWExpression rightSide, char operator) {
        this.leftSide = leftSide;
        this.rightSide = rightSide;
        this.operator = operator;
    }

    @Override
    public BWObject getValue() {
        return new BWInteger(value);
    }

    @Override
    public int getIntValue() {
        return value;
    }

    @Override
    public String getValueType() {
        return "integer";
    }

    @Override
    public void run(Scope scope) {
        leftSide.run(scope);
        int left = leftSide.getIntValue();
        rightSide.run(scope);
        int right = rightSide.getIntValue();
        switch (operator) {
            case '+':
                value = left + right;
                break;
            case '-':
                value = left - right;
                break;
            case '*':
                value = left * right;
                break;
            case '/':
                value = left / right;
                break;
            default:
                throw new IllegalArgumentException("Operator " + operator + " not known!");
        }
    }

    public char getOperator() {
        return operator;
    }

    public BWExpression getLeftSide() {
        return leftSide;
    }

    public BWExpression getRightSide() {
        return rightSide;
    }
}

File: src/main/java/org/dedda/bratwurst/lang/Calculation.kt
package org.dedda.bratwurst.lang

import org.dedda.bratwurst.lang.calculate.operationForChar
import org.dedda.bratwurst.lang.scope.Scope

/**
 * Created by dedda on 10/15/15.
 *
 * @author dedda
 */
class Calculation(lineNumber: Int, val leftSide: BWExpression, val rightSide: BWExpression, val operator: Char) :
    BWExpression(lineNumber) {

    private var value = 0
    override fun getValue(): BWObject {
        return BWInteger(value)
    }

    override fun getIntValue(): Int {
        return value
    }

    override fun getValueType(): ValueType {
        return ValueType.INTEGER
    }

    override fun run(scope: Scope) {
        leftSide.run(scope)
        val left = leftSide.intValue
        rightSide.run(scope)
        val right = rightSide.intValue
        value = operationForChar(operator)(left, right)
    }
}